Peer locator local addr

ivilata opened the pull request:
This change allows the helper to accept both the old ``PORT`` and the
new ``ADDRESS:PORT``, where ``ADDRESS`` is expected to be the local
outgoing interface address of the established connection as the peer
sees it, i.e. not as the helper sees it, since it may have been changed
by NAT.

The new information is saved locally as ``PUB_ADDR:PORT/?l=LOC_ADDR`` if
there's not an identical entry in the file.  A similarly formatted entry
may be provided back to the peer.  When the peer builds the URL (even if
it knows nothing about local addresses), the URL will reflect both the
public and local addresses of the other peer, which should help telling
apart connections not working because of NAT (a normal case) or because
of other reasons.

Old ``PUB_ADDR:PORT`` entries are kept and may also be returned to peers.

This PR builds upon #1.

ivilata committed:
get local address from peer (besides port), provide it back

This change allows the helper to accept both the old ``PORT`` and the new
``ADDRESS:PORT``, where ``ADDRESS`` is expected to be the local outgoing
interface address of the established connection as the peer sees it, i.e. not
as the helper sees it, since it may have been changed by NAT.

The new information is saved locally as ``PUB_ADDR:PORT/?l=LOC_ADDR`` if
there's not an identical entry in the file.  A similarly formatted entry may
be provided back to the peer.  When the peer builds the URL (even if it knows
nothing about local addresses), the URL will reflect both the public and local
addresses of the other peer, which should help telling apart connections not
working because of NAT (a normal case) or because of other reasons.

Old ``PUB_ADDR:PORT`` entries are kept and may also be returned to peers.

vmon reviewed (commented):
I would rather drop the entry if the local and the public ips are not the same. what do you think?

ivilata commented:
I considered the option of extending the spawned HTTP with support for UPnP, NAT-PMP or PCP to open a port in the router for NAT setups, so that we could also check support for those protocols.  However, programming support in Python either seems poor or overly complex, and AFAIK the quality of implementations in routers is very variable.  So maybe it's not worth it for the limited cases in which TCP NAT traversal will work.

The closest thing that I got in the Python world was to replace the simple HTTP server with some variation of [sendf](https://pypi.python.org/pypi/sendf/), but I didn't get it to work.  Maybe you know about some simple UPnP-enabled HTTP server out there?

Otherwise we can just drop HTTP servers behind NAT as you suggest.

ivilata commented:
My bad, I didn't look well enough.  Actually, using [miniupnpc](https://pypi.python.org/pypi/miniupnpc/1.9) is just a matter of a few lines, as shown [here](https://stackoverflow.com/a/35666064) and with more detail [here](https://github.com/miniupnp/miniupnp/blob/master/miniupnpc/testupnpigd.py).  I tried in a machine behind a normal home router and it worked without issues.

With that, we could extend the test (in combination with ``simple_http.py``) to check whether the host has a public IP address and if not, try to map the requested port in the router using UPnP first.  If it fails because the mapping is busy, the server fails as if the address was in use, otherwise (e.g. if UPnP is not supported by the gateway) it fails with a different code and the test stops trying to run a server.

Maybe this would increase the number of probes acting as servers.

ivilata commented:
I updated vmon/ooni-probe#4 to detect NAT and enable a temporary port mapping using UPnP in that case.

I think the new changes make the test more useful. What do you think?

ivilata committed:
do not keep remote peer local address

Just store whether it's using NAT or not, so that we don't leak local
addresses to other peers.

However, since connections to the helper travel in the clear, the local
address can still be sniffed.

ivilata committed:
receive NAT flag from peer, indicate with boolean query string arg

Thus peers do not need to share their local address.

New peers are distinguished from old ones because the former always send at
least one flag.

ivilata commented:
With the latest commits, peers no longer send their local address (for increased privacy, thus other peers or sniffers cannot see it). Instead, since the node now knows whether it sits behind NAT or not, it tells the helper, and the helper just stores a ``nat=(true|false)`` query argument which is transmitted to other peers.

The latest commits from vmon/ooni-probe#4 have the same purpose.
